webgl build appears to be broken

[421/442] Linking CXX executable web/filament-js/filament.js
FAILED: web/filament-js/filament.js 
: && /tmp/emsdk/upstream/emscripten/em++  -std=c++14 -fstrict-aliasing -Wno-unknown-pragmas -Wno-unused-function -fcolor-diagnostics  -DEMSCRIPTEN_HAS_UNBOUND_TYPE_NAMES=0 -DNDEBUG -O2 -fomit-frame-pointer -ffunction-sections -fdata-sections -fno-exceptions -fno-unwind-tables -fno-asynchronous-unwind-tables -fno-rtti  -s USE_WEBGL2=1  --bind -s ALLOW_MEMORY_GROWTH=1 -s MODULARIZE_INSTANCE=1 -s EXPORT_NAME=Filament -s ERROR_ON_UNDEFINED_SYMBOLS=0 --post-js /app/filament/web/filament-js/wasmloader.js --post-js /app/filament/web/filament-js/extensions.js --post-js /app/filament/web/filament-js/utilities.js -O2 web/filament-js/CMakeFiles/filament-js.dir/jsenums.cpp.o web/filament-js/CMakeFiles/filament-js.dir/jsbindings.cpp.o  -o web/filament-js/filament.js  filament/libfilament.a libs/math/libmath.a libs/utils/libutils.a libs/image/libimage.a libs/filameshio/libfilameshio.a libs/gltfio/libgltfio_core.a third_party/meshoptimizer/libmeshoptimizer.a filament/libfilament.a filament/backend/libbackend.a libs/filaflat/libfilaflat.a libs/filabridge/libfilabridge.a libs/ibl/libibl.a libs/geometry/libgeometry.a libs/math/libmath.a libs/utils/libutils.a libs/gltfio/libgltfio_resources.a third_party/draco/tnt/libdracodec.a && :
em++: error: Attempt to set a non-existent setting: 'MODULARIZE_INSTANCE'
 - did you mean one of MODULARIZE?
 - perhaps a typo in emcc's  -s X=Y  notation?
 - (see src/settings.js for valid values)
[427/442] Building CXX object third_party/imgui/tnt/CMakeFiles/imgui.dir/__/imgui.cpp.o
ninja: build stopped: subcommand failed.

--- a/BUILDING.md
+++ b/BUILDING.md
@@ -365,7 +365,7 @@ same version that our continuous builds use.
 
 ```
 cd <your chosen parent folder for the emscripten SDK>
-curl -L https://github.com/emscripten-core/emsdk/archive/1b1f08f.zip > emsdk.zip
+curl -L https://github.com/emscripten-core/emsdk/archive/f5e21de.zip > emsdk.zip
 unzip emsdk.zip ; mv emsdk-* emsdk ; cd emsdk
 python ./emsdk.py install latest
 python ./emsdk.py activate latest

--- a/build/web/ci-common.sh
+++ b/build/web/ci-common.sh
@@ -9,7 +9,7 @@ export PATH="$PWD:$PATH"
 # npm install -g typescript
 
 # Install emscripten.
-curl -L https://github.com/emscripten-core/emsdk/archive/1b1f08f.zip > emsdk.zip
+curl -L https://github.com/emscripten-core/emsdk/archive/f5e21de.zip > emsdk.zip
 unzip emsdk.zip ; mv emsdk-* emsdk ; cd emsdk
 ./emsdk install latest
 ./emsdk activate latest

--- a/web/filament-js/CMakeLists.txt
+++ b/web/filament-js/CMakeLists.txt
@@ -2,10 +2,13 @@ cmake_minimum_required(VERSION 3.10)
 
 project(filament-js)
 
-set(JAVASCRIPT_SRC
+set(EXTERN_POSTJS_SRC
   ${CMAKE_CURRENT_SOURCE_DIR}/wasmloader.js
   ${CMAKE_CURRENT_SOURCE_DIR}/extensions.js
   ${CMAKE_CURRENT_SOURCE_DIR}/utilities.js)
+
+set(POSTJS_SRC
+  ${CMAKE_CURRENT_SOURCE_DIR}/createContext.js)
 
 set(CPP_SRC
   jsenums.cpp
@@ -22,11 +25,15 @@ set(COPTS "${COPTS} -DEMSCRIPTEN_HAS_UNBOUND_TYPE_NAMES=0")
 
 set(LOPTS "${LOPTS} --bind")
 set(LOPTS "${LOPTS} -s ALLOW_MEMORY_GROWTH=1")
-set(LOPTS "${LOPTS} -s MODULARIZE_INSTANCE=1")
+set(LOPTS "${LOPTS} -s MODULARIZE=1")
 set(LOPTS "${LOPTS} -s EXPORT_NAME=Filament")
 set(LOPTS "${LOPTS} -s ERROR_ON_UNDEFINED_SYMBOLS=0")
-foreach (JS_FILENAME ${JAVASCRIPT_SRC})
-  set(LOPTS "${LOPTS} --post-js ${JS_FILENAME}")
+set(LOPTS "${LOPTS} -s MIN_WEBGL_VERSION=2")
+set(LOPTS "${LOPTS} -s MAX_WEBGL_VERSION=2")
+set(LOPTS "${LOPTS} --post-js ${POSTJS_SRC}")
+
+foreach (JS_FILENAME ${EXTERN_POSTJS_SRC})
+  set(LOPTS "${LOPTS} --extern-post-js ${JS_FILENAME}")
 endforeach()
 
 set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} ${LOPTS}")
@@ -35,7 +42,7 @@ set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${COPTS}")
 add_executable(filament-js ${CPP_SRC})
 
 set_target_properties(filament-js PROPERTIES
-    LINK_DEPENDS "${JAVASCRIPT_SRC}"
+    LINK_DEPENDS "${EXTERN_POSTJS_SRC};${POSTJS_SRC}"
     OUTPUT_NAME filament)
 
 target_link_libraries(filament-js PRIVATE filament math utils image filameshio gltfio_core stb)

--- a/web/filament-js/createContext.js
+++ b/web/filament-js/createContext.js
@@ -1,0 +1,6 @@
+Filament.createContext = (canvas, options) => {
+    const ctx = canvas.getContext("webgl2", options);
+    const handle = GL.registerContext(ctx, options);
+    GL.makeContextCurrent(handle);
+    return ctx;
+};

--- a/web/filament-js/extensions.js
+++ b/web/filament-js/extensions.js
@@ -45,9 +45,7 @@ Filament.loadClassExtensions = function() {
         options = Object.assign(defaults, options);
 
         // Create the WebGL 2.0 context and register it with emscripten.
-        const ctx = canvas.getContext("webgl2", options);
-        const handle = GL.registerContext(ctx, options);
-        GL.makeContextCurrent(handle);
+        const ctx = Filament.createContext(canvas, options);
 
         // Enable all desired extensions by calling getExtension on each one.
         ctx.getExtension('WEBGL_compressed_texture_s3tc');
@@ -315,7 +313,7 @@ Filament.loadClassExtensions = function() {
         // Construct the set of URI strings to fetch.
         const urlset = new Set();
         const urlToName = {};
-        for (var i = 0; i < names.size(); i++) {
+        for (let i = 0; i < names.size(); i++) {
             const name = names.get(i);
             if (name) {
                 const url = '' + new URL(name, basePath);

--- a/web/filament-js/jsbindings.cpp
+++ b/web/filament-js/jsbindings.cpp
@@ -987,7 +987,8 @@ class_<Material>("Material")
     .function("getDefaultInstance",
             select_overload<MaterialInstance*(void)>(&Material::getDefaultInstance),
             allow_raw_pointers())
-    .function("createInstance", &Material::createInstance, allow_raw_pointers())
+    .function("createInstance", EMBIND_LAMBDA(MaterialInstance*, (Material* self), {
+        return self->createInstance(); }), allow_raw_pointers())
     .function("createNamedInstance", EMBIND_LAMBDA(MaterialInstance*,
             (Material* self, std::string name), {
         return self->createInstance(name.c_str()); }), allow_raw_pointers())

--- a/web/filament-js/utilities.js
+++ b/web/filament-js/utilities.js
@@ -82,7 +82,7 @@ Filament.CompressedPixelBuffer = function(typedarray, cdatatype, faceSize) {
 Filament._loadFilamesh = function(engine, buffer, definstance, matinstances) {
     matinstances = matinstances || {};
     const registry = new Filament.MeshReader$MaterialRegistry();
-    for (var key in matinstances) {
+    for (let key in matinstances) {
         registry.set(key, matinstances[key]);
     }
     if (definstance) {
@@ -90,7 +90,7 @@ Filament._loadFilamesh = function(engine, buffer, definstance, matinstances) {
     }
     const mesh = Filament.MeshReader.loadMeshFromBuffer(engine, buffer, registry);
     const keys = registry.keys();
-    for (var i = 0; i < keys.size(); i++) {
+    for (let i = 0; i < keys.size(); i++) {
         const key = keys.get(i);
         const minstance = registry.get(key);
         matinstances[key] = minstance;
@@ -175,8 +175,8 @@ Filament.IcoSphere.prototype.subdivide = function() {
     const dsttris = new Uint16Array(ndsttris * 3);
     const dstverts = new Float32Array(ndstverts * 3);
     dstverts.set(srcverts);
-    var srcind = 0, dstind = 0, i3 = nsrcverts * 3, i4 = i3 + 3, i5 = i4 + 3;
-    for (var tri = 0; tri < nsrctris; tri++, i3 += 9, i4 += 9, i5 += 9) {
+    let srcind = 0, dstind = 0, i3 = nsrcverts * 3, i4 = i3 + 3, i5 = i4 + 3;
+    for (let tri = 0; tri < nsrctris; tri++, i3 += 9, i4 += 9, i5 += 9) {
         const i0 = srctris[srcind++] * 3;
         const i1 = srctris[srcind++] * 3;
         const i2 = srctris[srcind++] * 3;
@@ -288,7 +288,7 @@ Filament._createTextureFromImageFile = function(fileContents, engine, options) {
 
     const decodedImage = Filament.decodeImage(fileContents, noalpha ? 3 : 4);
 
-    var texformat, pbformat, pbtype;
+    let texformat, pbformat, pbtype;
     if (noalpha) {
         texformat = srgb ? TextureFormat.SRGB8 : TextureFormat.RGB8;
         pbformat = PixelDataFormat.RGB;
@@ -322,15 +322,15 @@ Filament.getSupportedFormats = function() {
         return Filament.supportedFormats;
     }
     const options = { majorVersion: 2, minorVersion: 0 };
-    var ctx = document.createElement('canvas').getContext('webgl2', options);
+    let ctx = document.createElement('canvas').getContext('webgl2', options);
     const result = {
         s3tc: false,
         astc: false,
         etc: false,
     }
-    var exts = ctx.getSupportedExtensions(), nexts = exts.length, i;
+    let exts = ctx.getSupportedExtensions(), nexts = exts.length, i;
     for (i = 0; i < nexts; i++) {
-        var ext = exts[i];
+        let ext = exts[i];
         if (ext == "WEBGL_compressed_texture_s3tc") {
             result.s3tc = true;
         } else if (ext == "WEBGL_compressed_texture_astc") {
@@ -350,8 +350,8 @@ Filament.getSupportedFormats = function() {
 /// ::retval:: empty string if there is no intersection of supported and desired formats.
 Filament.getSupportedFormatSuffix = function(desiredFormats) {
     desiredFormats = desiredFormats.split(' ');
-    var exts = Filament.getSupportedFormats();
-    for (var key in exts) {
+    let exts = Filament.getSupportedFormats();
+    for (let key in exts) {
         if (exts[key] && desiredFormats.includes(key)) {
             return '_' + key;
         }

--- a/web/filament-js/wasmloader.js
+++ b/web/filament-js/wasmloader.js
@@ -14,18 +14,9 @@
  * limitations under the License.
  */
 
- /*
- * WHY DOES THIS FILE EXIST?
- * -------------------------
- *
- * For each generated wasm file, emscripten generates a corollary JS file that is used to load the
- * wasm and set up its imports / exports. This file is a hand-authored addendum to the JS file that
- * emscripten generates, and we use it to store JavaScript-side utilities such as the asset
- * downloader and BufferDescriptor wrapper.
- *
- * Note that we should avoid certain modern JavaScript features in this file (e.g., for-of loops)
- * because emscripten uses a fairly rusty minifier.
- */
+// Emscripten creates a global function called "Filament" but our TypeScript bindings assume that
+// Filament is a namespace not a function. Here we replace the function with the namespace.
+Filament = Filament();
 
 // Keep a counter of remaining asynchronous tasks (e.g., downloading assets) that must occur before
 // allowing the app to initialize. As soon as the counter hits zero, we know it is time to call the
@@ -89,7 +80,7 @@ Filament.postRun = function() {
 /// onDone ::argument:: callback that gets invoked after all assets have been downloaded.
 /// onFetch ::argument:: optional callback that's invoked after each asset is downloaded.
 Filament.fetch = function(assets, onDone, onFetched) {
-    var remainingAssets = assets.length;
+    let remainingAssets = assets.length;
     assets.forEach(function(name) {
 
         // Check if a buffer already exists in case the client wishes

